### ISSUE TITLE
Multi-GPU Training Hangs (NCCL Timeout) with Dynamic Graph and find_unused_parameters

### ISSUE THREAD

**Issue by xiamengxxx** [Anonymous324-star/GeneMamba#3]
Hello,

First of all, thank you for your great work on this project! I'm currently trying to pre-train the model on a multi-GPU setup and have encountered a persistent issue that I'd like to ask about.

**Environment**
transformers version:4.40.2

torch version:2.3.0

mamba-ssm version:2.2.2

Python version:3.9.19

CUDA version:12.1

GPUs: 4 x H100

**Bug Description**
When running the pre-training script using torchrun with multiple GPUs, the training process hangs indefinitely at the very beginning of the first epoch (0% progress). After about 10 minutes, it crashes with an NCCL collective operation timeout error.

Single-GPU training works perfectly fine.

Error Log Snippet:

[rank0]:[E ProcessGroupNCCL.cpp:563] [Rank 0] Watchdog caught collective operation timeout: WorkNCCL(SeqNum=1, OpType=ALLREDUCE, ...) ran for 600072 milliseconds before timing out.
...
terminate called after throwing an instance of 'c10::DistBackendError'

**Analysis and Diagnosis**
After some debugging, it appears the root cause is a deadlock during distributed training, related to the custom MambaTrainer and its info_nce_loss calculation.

**Dynamic Computation Graph**: The compute_loss method uses a conditional check: if num_positive_pairs > 0:.

**Divergence Across GPUs**: In a multi-GPU setup (DDP), the batch of data on each GPU is different. It's possible for one GPU (e.g., Rank 0) to have positive pairs, thus executing the if block and generating gradients for the parameters involved in the InfoNCE loss. Simultaneously, another GPU (e.g., Rank 1) might get a batch with zero positive pairs, executing the else block and not generating gradients for those same parameters.

**Deadlock with find_unused_parameters=True**: With the default ddp_find_unused_parameters=True in TrainingArguments, DDP tries to optimize communication by only synchronizing gradients for parameters that were used. This creates a deadlock: Rank 0 waits for a gradient from Rank 1 that Rank 1 will never send, because from its perspective, the parameter was never used.

**My Questions for You**
Is this behavior something you have encountered during your own experiments?

Or, is there another intended method to handle this, perhaps through a specific data sampling strategy that ensures positive pairs in every batch?

Thank you again for your time and for sharing this valuable project with the community!
